File: taurus-protect-sdk-typescript/src/services/change-service.ts
/**
 * Change service for Taurus-PROTECT SDK.
 *
 * Provides methods for managing configuration changes that require approval.
 */

import { NotFoundError, ValidationError } from '../errors';
import type { ChangesApi } from '../internal/openapi/apis/ChangesApi';
import { changeFromDto, createChangeRequestToDto, listChangesResultFromDto } from '../mappers/change';
import type {
  Change,
  CreateChangeRequest,
  ListChangesOptions,
  ListChangesForApprovalOptions,
  ListChangesResult,
} from '../models/change';
import { BaseService } from './base';

/**
 * Service for managing configuration changes in the Taurus-PROTECT system.
 *
 * Changes represent modifications to system configuration that require approval
 * before taking effect. This includes user management, role assignments, and
 * other administrative operations that follow an approval workflow.
 *
 * @example
 * ```typescript
 * // Get changes pending approval
 * const result = await changeService.listForApproval({ pageSize: 50 });
 * for (const change of result.changes) {
 *   console.log(`${change.action} on ${change.entity} by ${change.creatorId}`);
 * }
 *
 * // Approve a change
 * await changeService.approve(changeId);
 *
 * // Reject a change
 * await changeService.reject(changeId);
 *
 * // Get changes with filters
 * const userChanges = await changeService.list({
 *   entity: 'user',
 *   status: 'Created',
 * });
 * ```
 */
export class ChangeService extends BaseService {
  private readonly changesApi: ChangesApi;

  /**
   * Creates a new ChangeService instance.
   *
   * @param changesApi - The ChangesApi instance from the OpenAPI client
   */
  constructor(changesApi: ChangesApi) {
    super();
    this.changesApi = changesApi;
  }

  /**
   * Gets a change by ID.
   *
   * @param id - The change ID
   * @returns The change
   * @throws {@link ValidationError} If id is empty
   * @throws {@link NotFoundError} If change is not found
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const change = await changeService.get('123');
   * console.log(`${change.action} on ${change.entity}`);
   * ```
   */
  async get(id: string): Promise<Change> {
    if (!id || id.trim() === '') {
      throw new ValidationError('id is required');
    }

    return this.execute(async () => {
      const response = await this.changesApi.changeServiceGetChange({ id });
      const resp = response as Record<string, unknown>;
      const result = resp.result ?? resp.change;
      const change = changeFromDto(result);

      if (!change) {
        throw new NotFoundError(`Change with id '${id}' not found`);
      }

      return change;
    });
  }

  /**
   * Lists changes with optional filtering.
   *
   * @param options - Optional filtering options
   * @returns The list of changes with pagination info
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * // List all changes
   * const result = await changeService.list();
   *
   * // Filter by entity type and status
   * const userChanges = await changeService.list({
   *   entity: 'user',
   *   status: 'Created',
   * });
   *
   * // Paginate through results
   * const firstPage = await changeService.list({ pageSize: 50 });
   * if (firstPage.hasNext) {
   *   const nextPage = await changeService.list({
   *     pageSize: 50,
   *     currentPage: firstPage.currentPage,
   *     pageRequest: 'NEXT',
   *   });
   * }
   * ```
   */
  async list(options?: ListChangesOptions): Promise<ListChangesResult> {
    return this.execute(async () => {
      const response = await this.changesApi.changeServiceGetChanges({
        entity: options?.entity,
        status: options?.status,
        creatorId: options?.creatorId,
        sortOrder: options?.sortOrder,
        cursorCurrentPage: options?.currentPage,
        cursorPageRequest: options?.pageRequest,
        cursorPageSize: options?.pageSize?.toString(),
        entityIDs: options?.entityIDs,
        entityUUIDs: options?.entityUUIDs,
      });

      return listChangesResultFromDto(response);
    });
  }

  /**
   * Lists changes pending approval.
   *
   * @param options - Optional filtering options
   * @returns The list of changes with pagination info
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * // Get all changes pending approval
   * const result = await changeService.listForApproval();
   *
   * // Filter by entity types
   * const userAndGroupChanges = await changeService.listForApproval({
   *   entities: ['user', 'group'],
   * });
   * ```
   */
  async listForApproval(options?: ListChangesForApprovalOptions): Promise<ListChangesResult> {
    return this.execute(async () => {
      const response = await this.changesApi.changeServiceGetChangesForApproval({
        entities: options?.entities,
        sortOrder: options?.sortOrder,
        cursorCurrentPage: options?.currentPage,
        cursorPageRequest: options?.pageRequest,
        cursorPageSize: options?.pageSize?.toString(),
        entityIDs: options?.entityIDs,
        entityUUIDs: options?.entityUUIDs,
      });

      return listChangesResultFromDto(response);
    });
  }

  /**
   * Approves a change.
   *
   * @param id - The change ID to approve
   * @throws {@link ValidationError} If id is empty
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * await changeService.approve('123');
   * ```
   */
  async approve(id: string): Promise<void> {
    if (!id || id.trim() === '') {
      throw new ValidationError('id is required');
    }

    return this.execute(async () => {
      await this.changesApi.changeServiceApproveChange({ id, body: {} });
    });
  }

  /**
   * Approves multiple changes.
   *
   * @param ids - The list of change IDs to approve
   * @throws {@link ValidationError} If ids is empty
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * await changeService.approveMany(['123', '456', '789']);
   * ```
   */
  async approveMany(ids: string[]): Promise<void> {
    if (!ids || ids.length === 0) {
      throw new ValidationError('ids cannot be empty');
    }

    return this.execute(async () => {
      await this.changesApi.changeServiceApproveChanges({
        body: { ids },
      });
    });
  }

  /**
   * Rejects a change.
   *
   * @param id - The change ID to reject
   * @throws {@link ValidationError} If id is empty
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * await changeService.reject('123');
   * ```
   */
  async reject(id: string): Promise<void> {
    if (!id || id.trim() === '') {
      throw new ValidationError('id is required');
    }

    return this.execute(async () => {
      await this.changesApi.changeServiceRejectChange({ id, body: {} });
    });
  }

  /**
   * Rejects multiple changes.
   *
   * @param ids - The list of change IDs to reject
   * @throws {@link ValidationError} If ids is empty
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * await changeService.rejectMany(['123', '456', '789']);
   * ```
   */
  async rejectMany(ids: string[]): Promise<void> {
    if (!ids || ids.length === 0) {
      throw new ValidationError('ids cannot be empty');
    }

    return this.execute(async () => {
      await this.changesApi.changeServiceRejectChanges({
        body: { ids },
      });
    });
  }

  /**
   * Creates a change request.
   *
   * @param request - The change request details
   * @returns The created change ID
   * @throws {@link ValidationError} If request is missing or has invalid fields
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const changeId = await changeService.create({
   *   action: 'update',
   *   entity: 'businessrule',
   *   entityId: '123',
   *   changes: { rulevalue: '100' },
   *   comment: 'Update rule value',
   * });
   * ```
   */
  async create(request: CreateChangeRequest): Promise<string> {
    if (!request) {
      throw new ValidationError('request is required');
    }
    if (!request.action || request.action.trim() === '') {
      throw new ValidationError('action is required');
    }
    if (!request.entity || request.entity.trim() === '') {
      throw new ValidationError('entity is required');
    }

    return this.execute(async () => {
      const dto = createChangeRequestToDto(request);
      const response = await this.changesApi.changeServiceCreateChange({ body: dto as any });
      const resp = response as Record<string, unknown>;
      const result = resp.result as Record<string, unknown> | undefined;
      return result?.id != null ? String(result.id) : '';
    });
  }
}

File: taurus-protect-sdk-python/taurus_protect/services/change_service.py
"""Change service for Taurus-PROTECT SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, List, Optional

from taurus_protect.mappers.audit import (
    change_from_dto,
    changes_from_dto,
    create_change_request_to_dto,
)
from taurus_protect.models.audit import (
    Change,
    ChangeResult,
    CreateChangeRequest,
    ListChangesOptions,
)
from taurus_protect.models.pagination import cursor_page
from taurus_protect.services._base import BaseService

if TYPE_CHECKING:
    pass  # For OpenAPI types when available


class ChangeService(BaseService):
    """Service for managing configuration changes.

    Changes represent modifications to system configuration that require
    approval before taking effect (dual-admin workflow).
    """

    def __init__(self, api_client: Any, changes_api: Any) -> None:
        super().__init__(api_client)
        self._changes_api = changes_api

    def create_change(self, request: CreateChangeRequest) -> str:
        """Create a change request.

        Args:
            request: The change request with action, entity, changes, etc.

        Returns:
            The created change ID.

        Raises:
            ValueError: If action or entity are empty.
            APIError: If API request fails.
        """
        if request is None:
            raise ValueError("request cannot be None")
        if not request.action or not request.action.strip():
            raise ValueError("action cannot be empty")
        if not request.entity or not request.entity.strip():
            raise ValueError("entity cannot be empty")

        try:
            from taurus_protect._internal.openapi.models.tgvalidatord_create_change_request import (
                TgvalidatordCreateChangeRequest,
            )

            dto_dict = create_change_request_to_dto(request)
            body = TgvalidatordCreateChangeRequest(**dto_dict)
            reply = self._changes_api.change_service_create_change(body)
            result = getattr(reply, "result", None)
            if result is None:
                from taurus_protect.errors import APIError

                raise APIError("createChange returned no result")
            return getattr(result, "id", "")
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            raise self._handle_error(e) from e

    def get(self, change_id: str) -> Change:
        """Get a change by ID.

        Args:
            change_id: The change ID to retrieve.

        Returns:
            The change.

        Raises:
            ValueError: If change_id is invalid.
            NotFoundError: If change not found.
            APIError: If API request fails.
        """
        self._validate_required(change_id, "change_id")

        try:
            resp = self._changes_api.change_service_get_change(change_id)

            result = getattr(resp, "result", None)
            if result is None:
                from taurus_protect.errors import NotFoundError

                raise NotFoundError(f"Change {change_id} not found")

            change = change_from_dto(result)
            if change is None:
                from taurus_protect.errors import NotFoundError

                raise NotFoundError(f"Change {change_id} not found")

            return change
        except Exception as e:
            from taurus_protect.errors import APIError, NotFoundError

            if isinstance(e, (APIError, NotFoundError, ValueError)):
                raise
            raise self._handle_error(e) from e

    def list(
        self,
        options: Optional[ListChangesOptions] = None,
    ) -> ChangeResult:
        """List changes, one page at a time.

        Args:
            options: Filters and page window; ``options.cursor`` continues a walk.

        Returns:
            ChangeResult with the changes and the page.

        Raises:
            ValueError: If the page size is invalid or cursor options conflict.
            APIError: If API request fails.
        """
        opts = options or ListChangesOptions()
        req = opts.to_cursor_request()

        try:
            resp = self._changes_api.change_service_get_changes(
                entity=opts.entity,
                entity_id=opts.entity_id,
                status=opts.status,
                creator_id=opts.creator_id,
                sort_order=opts.sort_order,
                entity_ids=opts.entity_ids,
                entity_uuids=opts.entity_uuids,
                **req.query_params(),
            )

            return ChangeResult(
                changes=changes_from_dto(resp.result),
                page=cursor_page(req.page_size, resp.cursor),
            )
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            raise self._handle_error(e) from e

    def list_for_approval(
        self,
        options: Optional[ListChangesOptions] = None,
    ) -> ChangeResult:
        """List changes pending approval, one page at a time.

        Args:
            options: Filters and page window. ``entity`` filters the queue by entity;
                ``status``, ``creator_id`` and ``entity_id`` do not apply to it and are
                refused.

        Returns:
            ChangeResult with the changes and the page.

        Raises:
            ValueError: If an option does not apply to the approval queue, the page size
                is invalid, or cursor options conflict.
            APIError: If API request fails.
        """
        opts = options or ListChangesOptions()
        for name in ("status", "creator_id", "entity_id"):
            if getattr(opts, name) is not None:
                raise ValueError(f"{name} cannot be applied to the change approval queue")
        req = opts.to_cursor_request()

        try:
            resp = self._changes_api.change_service_get_changes_for_approval(
                entities=[opts.entity] if opts.entity else None,
                sort_order=opts.sort_order,
                entity_ids=opts.entity_ids,
                entity_uuids=opts.entity_uuids,
                **req.query_params(),
            )

            return ChangeResult(
                changes=changes_from_dto(resp.result),
                page=cursor_page(req.page_size, resp.cursor),
            )
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            raise self._handle_error(e) from e

    def approve_change(self, change_id: str) -> None:
        """Approve a change.

        Args:
            change_id: The change ID to approve.

        Raises:
            ValueError: If change_id is invalid.
            APIError: If API request fails.
        """
        self._validate_required(change_id, "change_id")

        try:
            self._changes_api.change_service_approve_change(change_id, body={})
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            raise self._handle_error(e) from e

    def approve_changes(self, change_ids: List[str]) -> None:
        """Approve multiple changes.

        Args:
            change_ids: List of change IDs to approve.

        Raises:
            ValueError: If change_ids is empty or invalid.
            APIError: If API request fails.
        """
        if not change_ids:
            raise ValueError("change_ids cannot be empty")

        try:
            from taurus_protect._internal.openapi.models.tgvalidatord_approve_changes_request import (
                TgvalidatordApproveChangesRequest,
            )

            body = TgvalidatordApproveChangesRequest(ids=change_ids)
            self._changes_api.change_service_approve_changes(body)
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            raise self._handle_error(e) from e

    def reject_change(self, change_id: str) -> None:
        """Reject a change.

        Args:
            change_id: The change ID to reject.

        Raises:
            ValueError: If change_id is invalid.
            APIError: If API request fails.
        """
        self._validate_required(change_id, "change_id")

        try:
            self._changes_api.change_service_reject_change(change_id, body={})
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            raise self._handle_error(e) from e

    def reject_changes(self, change_ids: List[str]) -> None:
        """Reject multiple changes.

        Args:
            change_ids: List of change IDs to reject.

        Raises:
            ValueError: If change_ids is empty or invalid.
            APIError: If API request fails.
        """
        if not change_ids:
            raise ValueError("change_ids cannot be empty")

        try:
            from taurus_protect._internal.openapi.models.tgvalidatord_reject_changes_request import (
                TgvalidatordRejectChangesRequest,
            )

            body = TgvalidatordRejectChangesRequest(ids=change_ids)
            self._changes_api.change_service_reject_changes(body)
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            raise self._handle_error(e) from e
